BinaryInteger/advanced(by:) enhancement

oscbyspro opened the pull request:
#### Summary

1. Fixes an arbitrary, signed, binary integer crash case.
2. Improves the performance of small, unsigned, integers.

#### When does the current method crash?

`BinaryInteger/bitWidth` is a lower bound when `Self` is non-fixed-width:

```swift
:: BigInt(1).advanced(by: Int.max) 
-> BigInt(Int(truncatingIfNeeded: self) + Int.max) 
-> BigInt(Int(1) + Int.max) // crash
```

You can observe it with your favorite big integer package (here's mine: [Ultimathnum](https://github.com/oscbyspro/Ultimathnum)).

#### How does this improve small, unsigned, integer performance?

Small, unsigned, integers now take the same code path as small, signed, integers.

oscbyspro committed:
BinaryInteger/advanced(by:) enhancement.

1. Fixes an arbitrary, signed, binary integer crash case.
2. Improves the performance of small, unsigned, integers.

oscbyspro commented:
@stephentyrone, can you double-check whether this rewrite is sound?

stephentyrone commented:
@swift-ci benchmark

stephentyrone commented:
I'll take a look, also kicking off benchmarks to see what the code-size impact of this is in debug.

stephentyrone commented on stdlib/public/core/Integers.swift:
```diff
@@ -1501,14 +1501,20 @@ extension BinaryInteger {
+    : Self(exactly: UInt.max) == nil
+    
+    if typeIsSmall {
+      return Self(Int(truncatingIfNeeded: self) + n)
```
`FixedWidthInteger` imposes a requirement that the representable range be -2ⁿ ..< 2ⁿ, but `BinaryInteger` does not. So formally I think it's possible to have an unsigned `BinaryInteger` type `T` such that `T(exactly: UInt.max)` is nil, and yet `Int(truncatingIfNeeded: self)` can still wrap.

What do we lose if we define `typeIsSmall = Self(exactly: Int.max) == nil` unconditionally?

oscbyspro replied:
Hm, I see. In that case, truncating `UInt(bitPattern: 0 &- n)` would be bad if  `Int.max` fits but `Int.min.magnitude` doesn't. The result would be negative, however, so trapping through `Self.init(_:)` would solve that. 

Alternatively, `typeIsSmall` could be defined as `Self(exactly: Int.min.magnitude) == nil`. If an unsigned integer takes the small integer path, it is at most `Int.max`. Otherwise, the unsigned integer can represent all `Int` magnitudes. `Int` would take the small integer path, but same-type conversions should turn into no-ops.

What do you think about `typeIsSmall: Bool = Self(exactly: Int.min.magnitude) == nil` instead?

oscbyspro committed:
BinaryInteger/advanced(by:) enhancement (v2).

I have updated the small integer branch condition to `Self(exactly: Int.min.magnitude) == nil`. This is the optimal bound for unsigned integers since all small unsigned integers are bounded by `[0, Int.max]`. It works for (proper) binary signed integers since all small values are a subset of `Int` bounded by `[Int.min, Int.max]`. It also works for sign-magnitude integers since all small sign-magnitude values are a subset of `Int` bounded by `[-Int.max, Int.max]`. Any sign-magnitude integer larger than that can represent all `Int` values, so the non-small integer path also works.

oscbyspro commented:
I have updated the small integer branch condition to `Self(exactly: Int.min.magnitude) == nil`. This is the optimal bound for unsigned integers since all small unsigned integers are bounded by `[0, Int.max]`. It works for (proper) binary signed integers since all small values are a subset of `Int` bounded by `[Int.min, Int.max]`. It also works for sign-magnitude integers since all small sign-magnitude values are a subset of `Int` bounded by `[-Int.max, Int.max]`. Any sign-magnitude integer larger than that can represent all `Int` values, so the non-small integer path also works.

Note that it does not yet work for asymmetric, non-binary, signed integers. Imagine: `[2 * Int.min, 0]`. I would have considered such types invalid because there's no obvious solution for them, but I could think a bit harder about the problem if you want to support such conformances.

```swift
extension BinaryInteger {
  @inlinable
  @inline(__always)
  public func advanced(by n: Int) -> Self {
    // This compares its maximum bit width versus Int.bitWidth.
    // All small unsigned integers are bounded by [0, Int.max].
    // It should constant-fold since we use constant arguments.
    if Self(exactly: Int.min.magnitude) == nil {
      return Self(Int(truncatingIfNeeded: self) + n)
    } else if Self.isSigned {
      return self + Self(truncatingIfNeeded: n)
    } else if n < (0 as Int) {
      return self - Self(truncatingIfNeeded: UInt(bitPattern: 0 &- n))
    } else {
      return self + Self(truncatingIfNeeded: UInt(bitPattern: n))
    }
  }
}
```

stephentyrone commented:
I think I'm OK with this approach; the holes that it leaves are really very niche and unlikely to ever be a problem. I'll think about it a little bit more this morning, though.

oscbyspro commented:
After some more thought on this topic, I believe any signed `Swift.BinaryInteger` must be either unbounded or of the form `[-(x+1), x]`. This is because all signed `Swift.BinaryInteger` types must implement the bitwise `~` operator, which is `-(x+1)` in signed 2's complement terms. Only the aforementioned formats can return `-(x+1)` for all `x`.

stephentyrone commented:
Well, that's clearly false, since unsigned integers exist =)

oscbyspro commented:
Well, I did say signed integer ...thrice :D

stephentyrone commented:
If you're referring to types that conform to `FixedWidthInteger & SignedInteger`, then it's definitely true. It's less obvious that `BinaryInteger` enforces similar requirements (nor is it totally obvious that it doesn't--all of this could use some formalization). But in any event, I don't think I'm too worried about this particular corner case.

stephentyrone commented:
@Azoy does this help at all with the problem you were looking at?

stephentyrone commented:
@swift-ci test

stephentyrone commented:
macOS CI went belly-up before we even got to build swift, let's try again.

stephentyrone commented:
@swift-ci test macOS
